pkg/clients/clients.go

```diff
@@ -1,57 +1,163 @@
 package clients
 
 import (
+	"context"
 	"fmt"
 
 	"github.com/google/go-containerregistry/pkg/authn"
 	"github.com/openshift-psap/special-resource-operator/pkg/color"
 	buildv1 "github.com/openshift/api/build/v1"
+	configv1 "github.com/openshift/api/config/v1"
 	clientconfigv1 "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
 	"k8s.io/cli-runtime/pkg/genericclioptions"
 
+	v1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/rest"
+	restclient "k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/record"
+	controllerruntime "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 )
 
+//go:generate mockgen -source=clients.go -package=clients -destination=mock_clients_api.go
+
+const (
+	clusterVersionName = "version"
+)
+
 var (
-	log        = zap.New(zap.UseDevMode(true)).WithName(color.Print("clients", color.Brown))
-	Interface  *ClientsInterface
-	RestConfig *rest.Config
-	Namespace  string
-	config     genericclioptions.ConfigFlags
+	log       = zap.New(zap.UseDevMode(true)).WithName(color.Print("clients", color.Brown))
+	Interface ClientsInterface
+	// TODO need to remove this global variable
+	Namespace string
 )
 
-type ClientsInterface struct {
+type ClientsInterface interface {
```

Comment:
I think that this interface might be doing too many things at one. Its implementation, `k8sClients`, holds many Kubernetes clients. Shall we try to split it?

Reply:
Maybe another idea would be to try to refactor usage of this package? For example we have specific and generic ways of getting something: `Get`, `GetNamespace`, `ClusterOperatorGet`, `ClusterVersionGet`.

Reply:
We can use embedded interfaces:
```
ClusterOperator
ClusterVersion
...
```

Reply:
As discussed, we want this interface to hide the internal implementation from the other packages. So, a "user" package just gets/creates ClientsInterface and call the API, without needing to know which client is being used

Reply:
When having embedded interfaces you can call the "inner" methods directly from the most outer interface - this is just a matter of logic separation - for example in the future, we can re-use `ClusterOperatorClientInterface` interface without re-using the whole `ClientsInterface` interface.

Reply:
i don't think that using multiple interface is better: the "user" of the client just want to know how to execute some action on the cluster. They don't want to know the details. So why should a package care about if it needs one interface or another? This is the same annoying thing that we have with multiple kuberentes cluster, and every time you need to decide and understand which client to use for a specific API

Reply:
> So why should a package care about if it needs one interface or another?

He won't.

Maybe this code will explain better what I mean:
https://go.dev/play